### PR TITLE
Allow arrayElemAt to respect nested queries

### PR DESCRIPTION
**Pull request by msmerc**
Simple change to allow `$arrayElemAt` to work on things like `$foo.bar`.

Does this look ok?

### CODE REVIEW

File: core/src/test/java/de/bwaldvogel/mongo/backend/aggregation/ExpressionTest.java [hunk 1]
```diff
@@ -1265,7 +1274,7 @@ void testEvaluateStrLenCP() throws Exception {
         assertThat(Expression.evaluate(json("$strLenCP: '$a'"), json("a: 'value'"))).isEqualTo(5);
         assertThat(Expression.evaluate(json("$strLenCP: 'cafétéria'"), json(""))).isEqualTo(9);
         assertThat(Expression.evaluate(json("$strLenCP: '$a'"), json("a: '$€λA'"))).isEqualTo(4);
-        assertThat(Expression.evaluate(json("$strLenCP: '寿司'"), json(""))).isEqualTo(2);
+        assertThat(Expression.evaluate(json("$strLenCP: '\u5BFF\u53F8'"), json(""))).isEqualTo(2);
```
Comment (msmerc) [inline core/src/test/java/de/bwaldvogel/mongo/backend/aggregation/ExpressionTest.java:1277]: Windows is a bit funny about these characters. I've replaced them with their unicode equivalents.

Comment (bwaldvogel) [reply core/src/test/java/de/bwaldvogel/mongo/backend/aggregation/ExpressionTest.java:1277]: if it solves issues on Window, lets keep it as you suggested

**Comment by bwaldvogel** [comment]
Thanks for your contribution!

File: core/src/main/java/de/bwaldvogel/mongo/backend/aggregation/Expression.java [hunk 2]
```diff
@@ -163,7 +163,9 @@ Object apply(List<?> expressionValue, Document document) {
             if (index < 0 || index >= collection.size()) {
                 return null;
             } else {
-                return collection.get(index);
+                Object ret = collection.get(index);
+                if(Missing.isNullOrMissing(ret)) return null;
```
Comment (bwaldvogel) [inline core/src/main/java/de/bwaldvogel/mongo/backend/aggregation/Expression.java:167]: Actually the handling of "missing" and null values was not quite correct.
I’ve added an integration test in d44ca92b208def9e74f103d7e59bf17e388fff18 which compares the behavior to a real MongoDB.
I’ve adjusted the implementation to behave exactly like MongoDB does.

Comment (msmerc) [reply core/src/main/java/de/bwaldvogel/mongo/backend/aggregation/Expression.java:167]: Oh right- must confess I didn't test this edge case vs mongo, I just assumed it had to work like the existing tests 😮 

[I wonder if one could construct a test framework that runs the very same tests vs an actual mongo to check it's behaving the same?]

Comment (bwaldvogel) [reply core/src/main/java/de/bwaldvogel/mongo/backend/aggregation/Expression.java:167]: > I wonder if one could construct a test framework that runs the very same tests vs an actual mongo to check it's behaving the same?

That’s exactly what we already do :)

- `AbstractBackendTest` has multiple derived classes:
    1. `RealMongoBackendTest`
    2. `MemoryBackendTest`

Basically the same for `AbstractAggregationTest`
1. `RealMongoAggregationTest`
2. `MemoryBackendAggregationTest`

**Comment by bwaldvogel** [comment]
Version `1.39.0` contains the changes.
